File: src/main/java/frc/robot/subsystems/GrabberSubsystem.java
package frc.robot.subsystems;

import edu.wpi.first.wpilibj.DoubleSolenoid;
import edu.wpi.first.wpilibj.DoubleSolenoid.Value;
import edu.wpi.first.wpilibj2.command.Command;
import edu.wpi.first.wpilibj2.command.Commands;
import edu.wpi.first.wpilibj2.command.SubsystemBase;

public class GrabberSubsystem extends SubsystemBase {

  private final DoubleSolenoid leftSolenoid;
  private final DoubleSolenoid rightSolenoid;

  public GrabberSubsystem(DoubleSolenoid leftSolenoid, DoubleSolenoid rightSolenoid) {
    this.leftSolenoid = leftSolenoid;
    this.rightSolenoid = rightSolenoid;
    leftSolenoid.set(Value.kForward);
    rightSolenoid.set(Value.kForward);
  }

  public Command release() {
    return runOnce(this::openGrabber).andThen(Commands.waitSeconds(0.2));
  }

  public Command grab() {
    return runOnce(this::closeGrabber).andThen(Commands.waitSeconds(0.2));
  }

  public Command toggle() {
    return runOnce(this::toggleGrabber).andThen(Commands.waitSeconds(0.2));
  }

  private void closeGrabber() {
    leftSolenoid.set(Value.kReverse);
    rightSolenoid.set(Value.kReverse);
  }

  private void openGrabber() {
    leftSolenoid.set(Value.kForward);
    rightSolenoid.set(Value.kForward);
  }

  private void toggleGrabber() {
    leftSolenoid.toggle();
    rightSolenoid.toggle();
  }

  public boolean isOpen() {
    return leftSolenoid.get() == Value.kForward;
  }
}

File: src/main/java/frc/robot/subsystems/GrabberSubsystem.kt
package frc.robot.subsystems

import edu.wpi.first.wpilibj.DoubleSolenoid
import edu.wpi.first.wpilibj2.command.Command
import edu.wpi.first.wpilibj2.command.Commands
import edu.wpi.first.wpilibj2.command.SubsystemBase

class GrabberSubsystem(
    private val leftSolenoid: DoubleSolenoid,
    private val rightSolenoid: DoubleSolenoid
) : SubsystemBase() {
    init {
        openGrabber()
    }

    val release: Command
        get() = runOnce { openGrabber() }.andThen(Commands.waitSeconds(0.2))

    val grab: Command
        get() = runOnce { closeGrabber() }.andThen(Commands.waitSeconds(0.2))

    val toggle: Command
        get() = runOnce { toggleGrabber() }.andThen(Commands.waitSeconds(0.2))

    private fun closeGrabber() {
        leftSolenoid.set(DoubleSolenoid.Value.kReverse)
        rightSolenoid.set(DoubleSolenoid.Value.kReverse)
    }

    private fun openGrabber() {
        leftSolenoid.set(DoubleSolenoid.Value.kForward)
        rightSolenoid.set(DoubleSolenoid.Value.kForward)
    }

    private fun toggleGrabber() {
        leftSolenoid.toggle()
        rightSolenoid.toggle()
    }

    val isOpen: Boolean
        get() = leftSolenoid.get() == DoubleSolenoid.Value.kForward
}
